Additional fix for compile issue on OS X 10.10

crazyorc opened the pull request:
When building ROS on Mac OS X (yosemite), rosbag_storage and roscpp have compile errors (`error: no type named 'allocator' in namespace 'std'`).  The fix in issue #15 is extended to handle the packages which don't include any file which would cause either `__GLIBCXX__` or `_LIBCPP_VERSION` to be defined.

See: http://stackoverflow.com/a/25892335

crazyorc committed:
Additional fix for compile issue on OS X 10.10

When building ROS on Mac OS X (yosemite), rosbag_storage and roscpp have compile errors (error: no type named 'allocator' in namespace 'std').  The fix in issue #15 is extended to handle the packages which don't include any file which would cause either __GLIBCXX__ or _LIBCPP_VERSION to be defined.

See: http://stackoverflow.com/a/25892335

wjwwood commented:
What version of CLT's are you using? I'll try to reproduce based on that.

crazyorc commented:
I saw the issue with both CLT's 6.3 and 6.3.1.  Also, if it matters, I started from a clean Yosemite OS installation (eg. not upgrade from prior version), applied all the updates, and followed wiki instructions.

I downgraded to 6.2 once and tried that.  I'm suspicious of those results though as I didn't use xcode-select to switch to use the CLT's.  However, after that, I wiped and reinstalled Yosemite.

wjwwood commented:
> I downgraded to 6.2 once and tried that.

Suspicious or not, did it work? There was a bug in 6.3 which is fixed by either downgrading or creating a missing header, but it is seemingly unrelated to this problem:

http://stackoverflow.com/questions/29529455/missing-c-header-debug-after-updating-osx-command-line-tools-6-3/29576048#29576048

Anyways, I'll try to reproduce it when I get a chance.

crazyorc commented:
It did not work.  That's the SO question that started me down my troubleshooting path.  I tried to manually fix the issue by copying the missing `__debug` header from Xcode tools into the CLT and still got the same compile errors.

wjwwood commented:
Oh, well I was able to work around that with the solution in the above linked SO question. I used this:

```
$ echo '#define _LIBCPP_ASSERT(x, m) ((void)0)' | sudo tee -a /Library/Developer/CommandLineTools/usr/include/c++/v1/__debug > /dev/null
```

crazyorc commented:
I tried that too and still had the same compile errors.

FWIW, during my triage / troubleshooting, with each change (header, tools, ...) I did the following (until a build succeeded):
1) Ran build again
2) Removed build, devel, and install dirs.  Ran build again
3) Removed build, devel, install, and src dirs.  Ran build again starting with deps.
4) Removed build, devel, install, and src dirs.  Removed .rosinstall file. Regenerated .rosinstall.  Ran build again starting with deps.

Once I found the fix in this pull request, I reverted all the headers except that one.  Then I removed build, devel, and install folders and ran build again.  The goal being to see if it compiled in multiple scenarios issue free before committing and making pull request.

EDIT: I got the notion about `<memory>` include from http://en.cppreference.com/w/cpp/memory/allocator.  That set me to looking where that particular header was included, which led me to the `#ifndef _LIBCPP_VERSION` check in message_forward.h.  From there I added a `#warning` to the else and discovered the "if" condition seemed to be true.

wjwwood commented:
I needed this fix when building jade on my machine just now too, so @dirk-thomas I would recommend we merge and release this. Though we don't have to do that before the jade release.

huangzonghao commented:
thank you!
problem solved!
